New guide | Adding custom DoH resolvers to DNSCloak (#16)
privacytoolsIO/blog.privacytools.io: collections/_posts/2019-08-24-adding-custom-doh-resolvers-to-dnscloak.md

---
title: Adding Custom DNS over HTTPS Resolvers to DNSCloak

> author: nitrohorse

Reviewer: ```suggestion
author: nitrohorse
editor: Jonah
```
😜

> layout: post

Author: Unsure if a cover image is required and if so, what it would be..

Reviewer: Definitely looks better with a cover photo (~~950x450~~ **1280x640** res), but I'll make one tomorrow for you if you want.

Author: Oh that would be great! :pray:

Author: Any update on this?

tags:
  - guides

>   - dns

Reviewer: ```suggestion
  - dns
  - dnscrypt
```

  - ios
---

> **[DNSCloak](https://apps.apple.com/us/app/dnscloak-secure-dns-client/id1452162351)** is an [open-source](https://github.com/s-s/dnscloak) DNSCrypt and DNS over HTTPS (DoH) client for iOS; giving users the ability to encrypt their DNS requests. While highly configurable, its user interface can be unintuitive and doesn't easily allow users to add custom DoH resolvers apart from the default ["public-resolvers" list that DNSCrypt provides](https://github.com/DNSCrypt/dnscrypt-resolvers/blob/master/v2/public-resolvers.md).

Author: Probably could expand a bit more but briefly on what DoH and DNSCrypt is and why they're important.

> **[DNSCloak](https://apps.apple.com/us/app/dnscloak-secure-dns-client/id1452162351)** is an [open-source](https://github.com/s-s/dnscloak) DNSCrypt and DNS over HTTPS (DoH) client for iOS; giving users the ability to encrypt their DNS requests. While highly configurable, its user interface can be unintuitive and doesn't easily allow users to add custom DoH resolvers apart from the default ["public-resolvers" list that DNSCrypt provides](https://github.com/DNSCrypt/dnscrypt-resolvers/blob/master/v2/public-resolvers.md).

Reviewer: I am *assuming* it uses a VPN profile...
```suggestion
**[DNSCloak](https://apps.apple.com/us/app/dnscloak-secure-dns-client/id1452162351)** is an [open-source](https://github.com/s-s/dnscloak) DNSCrypt and DNS over HTTPS (DoH) client for iOS, which gives users the ability to encrypt their DNS requests through the use of an on-device VPN profile. While highly configurable, its user interface can be unintuitive and doesn't easily allow users to add custom DoH resolvers apart from the default ["public-resolvers" list](https://github.com/DNSCrypt/dnscrypt-resolvers/blob/master/v2/public-resolvers.md) that the DNSCrypt project provides.
```

Author: Yeah, you're right; good callout :+1:

> This guide will show how to use DNSCloak to connect to your favorite public resolver that supports DNSCrypt or DoH.

Reviewer: ```suggestion
This guide will walk you through setting up DNSCloak to connect to any public resolver that supports either DNSCrypt or DoH.
```

### Adding a Custom Resolver

> You'll notice DNSCloak provides a "Config Editor":

Reviewer: ```suggestion
DNSCloak provides a "Config Editor" which allows you to modify various aspects of the program, add custom servers, etc:
```

![](/assets/img/2019-08-24-dnscloak/config-editor.jpeg){: .w-50}

> We can learn about the various configuration options from the [example config file](https://github.com/jedisct1/dnscrypt-proxy/blob/master/dnscrypt-proxy/example-dnscrypt-proxy.toml) in dnscrypt-proxy's code repository. If you scroll all the way to the bottom, you'll find a `[static.'myserver']` section along with a `stamp` property. This stamp is for adding the [DNS stamp](https://dnscrypt.info/stamps-specifications) of your resolver.

Reviewer: ```suggestion
You can learn more about the various configuration options from the [example configuration file](https://github.com/jedisct1/dnscrypt-proxy/blob/master/dnscrypt-proxy/example-dnscrypt-proxy.toml) in dnscrypt-proxy's code repository. But, if you scroll all the way to the bottom you'll find a `[static.'myserver']` section along with a `stamp` property. This stamp is for adding the [DNS stamp](https://dnscrypt.info/stamps-specifications) of your resolver.
```

> > Server stamps encode all the parameters required to connect to a secure DNS server as a single string. Think about stamps as QR code, but for DNS.

Reviewer: ```suggestion
According to the DNSCrypt project, DNS stamps encode all the parameters required to connect to an encrypted DNS server as a single string. You can think about stamps as QR code, but for DNS.
```

I would probably not directly quote (partially because I am now realizing the blog doesn't have block quote support yet) and I would move this sentence to the end of the prior paragraph rather than have it on a separate line.

> At the time of writing this post, CZ.NIC's DoH resolver is the only one [PrivacyTools.io suggests](https://www.privacytools.io/providers/dns/#icanndns) that doesn't provide their users with a DNS stamp on their website, making adoption a bit more difficult. But thankfully, we can create a DNS stamp ourselves.

Reviewer: We haven't really officially decided but I think generally we refer to the site as `privacytools.io` and not `PrivacyTools.io`, `PTIO` or `Privacy Tools`? At least this is what I've been doing and have been led to believe.

```suggestion
### Generating a Stamp

Some providers will provide you with a DNS stamp pre-made for you. If your provider does this, great! You can skip ahead to the next section. At the time of writing this post, CZ.NIC's DoH resolver is the only provider [privacytools.io suggests](https://www.privacytools.io/providers/dns/#icanndns) that doesn't provide their users with a DNS stamp on their website, making adoption a bit more difficult. Thankfully however, we can create a DNS stamp ourselves.
```

We could also change "privacytools.io suggests" to "we suggest" if it sounds better?

Author: Definitely makes sense; will update to "we suggest" I think.

> DNSCrypt hosts a [DNS stamp calculator](https://dnscrypt.info/stamps/) (which you can also [download, compile, and run offline](https://github.com/jedisct1/vue-dnsstamp)) that we can fill out with the info CZ.NIC provides on their webpage to generate our stamp.

Reviewer: ```suggestion
To generate a DNS stamp, DNSCrypt hosts a [DNS stamp calculator](https://dnscrypt.info/stamps/) (which you can also [download, compile, and run offline](https://github.com/jedisct1/vue-dnsstamp)) that we can fill out with the information from our DNS provider. We'll be using CZ.NIC's information as an example to generate our stamp.
```

> We'll need three things about the DoH resolver:

Reviewer: ```suggestion
We will need to know three things about the DoH resolver you choose:
```

DoH only? This should also work with DNSCrypt right?

Author: Yeah, I'm specifically walking through DoH resolvers because I haven't personally done this with DNSCrypt resolvers -- the stamp requirements are also different in that they need:
- IP address
- "Provider public key"
- and a "Provider name"

1. IP address
2. Host name
3. Path

> Navigate to [CZ.NIC's webpage](https://www.nic.cz/odvr/) and scroll down to "How to turn on DNS-over-HTTPS (DoH)" to note the URL (https://odvr.nic.cz/doh).

Reviewer: ```suggestion
Browse to [CZ.NIC's webpage](https://www.nic.cz/odvr/)—there is an English language option at the top of the page—and scroll down to "How to turn on DNS-over-HTTPS (DoH)" and note the URL (in this case, `https://odvr.nic.cz/doh`).
```

Whenever I hear Navigate I think of Netscape 😆

![](/assets/img/2019-08-24-dnscloak/cz-nic-doh.png){: .w-100}

> Next, scroll up to the setup sections for either Windows, macOS, or Linux and copy one of the IPv4 addresses of the DoH resolver (`193.17.47.1` or `185.43.135.1`).

Reviewer: ```suggestion
Next, find one of the IPv4 addresses of the DoH resolver in any of the Windows, macOS, or Linux setup sections, and copy one of them (in this case, `193.17.47.1` or `185.43.135.1`).
```

> ![](/assets/img/2019-08-24-dnscloak/cz-nic-ips.png){: .w-50}

Reviewer: Question: do you think images that aren't `w-100` should be centered or are they fine as-is?

Author: Maybe centered would make the layout more readable I think.

Reviewer: Yeah, I'll work on that. I think that would actually make my [other idea](https://github.com/privacytoolsIO/blog.privacytools.io/pull/16/files#r317380068) easier to implement as well.

Now we can paste what we've gathered into the stamp calculator:
- IP address: `193.17.47.1`
- Host name: `odvr.nic.cz`
- Path: `/doh`

![](/assets/img/2019-08-24-dnscloak/cz-nic-stamp.png){: .w-100}

We'll find that the DNS stamp is `sdns://AgMAAAAAAAAACzE5My4xNy40Ny4xAAtvZHZyLm5pYy5jegQvZG9o`.

> Let's copy and paste our new configuration into DNSCloak's Config Editor at the bottom:

Reviewer: ```suggestion
### Adding Resolvers to DNSCloak

Now that we have a DNS stamp generated, we can copy and paste our new configuration into the bottom of DNSCloak's Config Editor, like so:
```

```
[static.'CZ.NIC-193.17.47.1']
stamp = 'sdns://AgMAAAAAAAAACzE5My4xNy40Ny4xAAtvZHZyLm5pYy5jegQvZG9o'
```

![](/assets/img/2019-08-24-dnscloak/config-editor-cz-nic.jpeg){: .w-50}

> Save by selecting the checkmark icon on the top right, and now we can use our new configuration:

Reviewer: ```suggestion
Select the checkmark icon in the top right corner to save your configuration, and it should be good to go!
```

![](/assets/img/2019-08-24-dnscloak/dnscloak-cz-nic.jpeg){: .w-50}

> Connect, and let's finally validate DNSCloak is working as expected by visiting [DNSLeakTest.com](https://dnsleaktest.com/):

Reviewer: ```suggestion
Get connected, and we can finally validate DNSCloak is working as expected by visiting [DNSLeakTest.com](https://dnsleaktest.com/):
```

![](/assets/img/2019-08-24-dnscloak/dnsleaktest-cz-nic.jpeg){: .w-100}

### Adding Cloudflare's Resolver for Firefox

Another public DoH resolver that we may want to use is Cloudflare's public resolver for _Firefox_ [which has a stricter logging policy than Cloudflare's default resolver](https://forum.privacytools.io/t/logging-differences-between-cloudflares-default-dns-over-https-resolver-and-their-resolver-for-firefox/1451).

We can [generate a stamp](https://dnscrypt.info/stamps/) with this information:
- IP address: `1.1.1.1`
- Host name: `mozilla.cloudflare-dns.com`
- Path: `/dns-query`

![](/assets/img/2019-08-24-dnscloak/cloudflare-mozilla-stamp.png){: .w-100}

> You can now paste this into DNSCloak's Config Editor and start using the resolver.

Reviewer: ```suggestion
You can now paste the following stamp we generated into DNSCloak's Config Editor and start using the resolver.
```

```
[static.'Cloudflare Resolver for Firefox']
stamp = 'sdns://AgUAAAAAAAAABzEuMS4xLjEAGm1vemlsbGEuY2xvdWRmbGFyZS1kbnMuY29tCi9kbnMtcXVlcnk'
```

![](/assets/img/2019-08-24-dnscloak/config-editor-cf-moz.jpeg){: .w-50}

> ![](/assets/img/2019-08-24-dnscloak/dnscloak-cf-moz.jpeg){: .w-50}

Author: Unsure if this needs some sort of conclusion :grimacing: Yeah, it does probably.

> ![](/assets/img/2019-08-24-dnscloak/dnscloak-cf-moz.jpeg){: .w-50}

Reviewer: I want to figure out how to get these pictures side-by-side...

Author: Oh that would be sweet

Author: Any update on this?
